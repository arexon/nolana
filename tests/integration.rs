use std::{fs, path::Path};

use insta::Settings;
use nolana::Parser;

fn with_settings(f: impl FnOnce()) {
    let mut settings = Settings::clone_current();
    settings.set_omit_expression(true);
    settings.set_prepend_module_to_snapshot(false);
    settings.bind(f)
}

fn read_and_parse(path: &Path) -> String {
    let source = fs::read_to_string(path).unwrap();
    let result = Parser::new(&source).parse();
    format!("{result:#?}")
}

#[test]
fn test_parser() {
    with_settings(|| {
        insta::glob!("parser/*.nolana", |path| {
            insta::assert_snapshot!(read_and_parse(path));
        });
    });
}
