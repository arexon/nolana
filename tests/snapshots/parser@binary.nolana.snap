---
source: tests/integration.rs
input_file: tests/parser/binary.nolana
---
ParseResult {
    program: Program {
        span: Span {
            start: 0,
            end: 85,
        },
        source: "1 + 2 * 3;\n1 >> 2 << 3 & 4 | 5 ^ 6;\n(1 + 1) * (1 + 1);\n((2 * 3) + 1) / 2;\nv.a ?? 1.2;\n",
        body: Complex(
            [
                Expression(
                    Binary(
                        BinaryExpression {
                            span: Span {
                                start: 0,
                                end: 9,
                            },
                            left: NumericLiteral(
                                NumericLiteral {
                                    span: Span {
                                        start: 0,
                                        end: 1,
                                    },
                                    value: 1.0,
                                    raw: "1",
                                },
                            ),
                            operator: Addition,
                            right: Binary(
                                BinaryExpression {
                                    span: Span {
                                        start: 4,
                                        end: 9,
                                    },
                                    left: NumericLiteral(
                                        NumericLiteral {
                                            span: Span {
                                                start: 4,
                                                end: 5,
                                            },
                                            value: 2.0,
                                            raw: "2",
                                        },
                                    ),
                                    operator: Multiplication,
                                    right: NumericLiteral(
                                        NumericLiteral {
                                            span: Span {
                                                start: 8,
                                                end: 9,
                                            },
                                            value: 3.0,
                                            raw: "3",
                                        },
                                    ),
                                },
                            ),
                        },
                    ),
                ),
                Expression(
                    Binary(
                        BinaryExpression {
                            span: Span {
                                start: 11,
                                end: 34,
                            },
                            left: Binary(
                                BinaryExpression {
                                    span: Span {
                                        start: 11,
                                        end: 26,
                                    },
                                    left: Binary(
                                        BinaryExpression {
                                            span: Span {
                                                start: 11,
                                                end: 22,
                                            },
                                            left: Binary(
                                                BinaryExpression {
                                                    span: Span {
                                                        start: 11,
                                                        end: 17,
                                                    },
                                                    left: NumericLiteral(
                                                        NumericLiteral {
                                                            span: Span {
                                                                start: 11,
                                                                end: 12,
                                                            },
                                                            value: 1.0,
                                                            raw: "1",
                                                        },
                                                    ),
                                                    operator: ShiftRight,
                                                    right: NumericLiteral(
                                                        NumericLiteral {
                                                            span: Span {
                                                                start: 16,
                                                                end: 17,
                                                            },
                                                            value: 2.0,
                                                            raw: "2",
                                                        },
                                                    ),
                                                },
                                            ),
                                            operator: ShiftLeft,
                                            right: NumericLiteral(
                                                NumericLiteral {
                                                    span: Span {
                                                        start: 21,
                                                        end: 22,
                                                    },
                                                    value: 3.0,
                                                    raw: "3",
                                                },
                                            ),
                                        },
                                    ),
                                    operator: BitwiseAnd,
                                    right: NumericLiteral(
                                        NumericLiteral {
                                            span: Span {
                                                start: 25,
                                                end: 26,
                                            },
                                            value: 4.0,
                                            raw: "4",
                                        },
                                    ),
                                },
                            ),
                            operator: BitwiseOr,
                            right: Binary(
                                BinaryExpression {
                                    span: Span {
                                        start: 29,
                                        end: 34,
                                    },
                                    left: NumericLiteral(
                                        NumericLiteral {
                                            span: Span {
                                                start: 29,
                                                end: 30,
                                            },
                                            value: 5.0,
                                            raw: "5",
                                        },
                                    ),
                                    operator: BitwiseXor,
                                    right: NumericLiteral(
                                        NumericLiteral {
                                            span: Span {
                                                start: 33,
                                                end: 34,
                                            },
                                            value: 6.0,
                                            raw: "6",
                                        },
                                    ),
                                },
                            ),
                        },
                    ),
                ),
                Expression(
                    Binary(
                        BinaryExpression {
                            span: Span {
                                start: 36,
                                end: 53,
                            },
                            left: Parenthesized(
                                ParenthesizedExpression {
                                    span: Span {
                                        start: 36,
                                        end: 43,
                                    },
                                    body: Single(
                                        Binary(
                                            BinaryExpression {
                                                span: Span {
                                                    start: 37,
                                                    end: 42,
                                                },
                                                left: NumericLiteral(
                                                    NumericLiteral {
                                                        span: Span {
                                                            start: 37,
                                                            end: 38,
                                                        },
                                                        value: 1.0,
                                                        raw: "1",
                                                    },
                                                ),
                                                operator: Addition,
                                                right: NumericLiteral(
                                                    NumericLiteral {
                                                        span: Span {
                                                            start: 41,
                                                            end: 42,
                                                        },
                                                        value: 1.0,
                                                        raw: "1",
                                                    },
                                                ),
                                            },
                                        ),
                                    ),
                                },
                            ),
                            operator: Multiplication,
                            right: Parenthesized(
                                ParenthesizedExpression {
                                    span: Span {
                                        start: 46,
                                        end: 53,
                                    },
                                    body: Single(
                                        Binary(
                                            BinaryExpression {
                                                span: Span {
                                                    start: 47,
                                                    end: 52,
                                                },
                                                left: NumericLiteral(
                                                    NumericLiteral {
                                                        span: Span {
                                                            start: 47,
                                                            end: 48,
                                                        },
                                                        value: 1.0,
                                                        raw: "1",
                                                    },
                                                ),
                                                operator: Addition,
                                                right: NumericLiteral(
                                                    NumericLiteral {
                                                        span: Span {
                                                            start: 51,
                                                            end: 52,
                                                        },
                                                        value: 1.0,
                                                        raw: "1",
                                                    },
                                                ),
                                            },
                                        ),
                                    ),
                                },
                            ),
                        },
                    ),
                ),
                Expression(
                    Binary(
                        BinaryExpression {
                            span: Span {
                                start: 55,
                                end: 72,
                            },
                            left: Parenthesized(
                                ParenthesizedExpression {
                                    span: Span {
                                        start: 55,
                                        end: 68,
                                    },
                                    body: Single(
                                        Binary(
                                            BinaryExpression {
                                                span: Span {
                                                    start: 56,
                                                    end: 67,
                                                },
                                                left: Parenthesized(
                                                    ParenthesizedExpression {
                                                        span: Span {
                                                            start: 56,
                                                            end: 63,
                                                        },
                                                        body: Single(
                                                            Binary(
                                                                BinaryExpression {
                                                                    span: Span {
                                                                        start: 57,
                                                                        end: 62,
                                                                    },
                                                                    left: NumericLiteral(
                                                                        NumericLiteral {
                                                                            span: Span {
                                                                                start: 57,
                                                                                end: 58,
                                                                            },
                                                                            value: 2.0,
                                                                            raw: "2",
                                                                        },
                                                                    ),
                                                                    operator: Multiplication,
                                                                    right: NumericLiteral(
                                                                        NumericLiteral {
                                                                            span: Span {
                                                                                start: 61,
                                                                                end: 62,
                                                                            },
                                                                            value: 3.0,
                                                                            raw: "3",
                                                                        },
                                                                    ),
                                                                },
                                                            ),
                                                        ),
                                                    },
                                                ),
                                                operator: Addition,
                                                right: NumericLiteral(
                                                    NumericLiteral {
                                                        span: Span {
                                                            start: 66,
                                                            end: 67,
                                                        },
                                                        value: 1.0,
                                                        raw: "1",
                                                    },
                                                ),
                                            },
                                        ),
                                    ),
                                },
                            ),
                            operator: Division,
                            right: NumericLiteral(
                                NumericLiteral {
                                    span: Span {
                                        start: 71,
                                        end: 72,
                                    },
                                    value: 2.0,
                                    raw: "2",
                                },
                            ),
                        },
                    ),
                ),
                Expression(
                    Binary(
                        BinaryExpression {
                            span: Span {
                                start: 74,
                                end: 84,
                            },
                            left: Variable(
                                VariableExpression {
                                    span: Span {
                                        start: 74,
                                        end: 77,
                                    },
                                    lifetime: Variable,
                                    member: Property {
                                        property: Identifier {
                                            span: Span {
                                                start: 76,
                                                end: 77,
                                            },
                                            name: "a",
                                        },
                                    },
                                },
                            ),
                            operator: Coalesce,
                            right: NumericLiteral(
                                NumericLiteral {
                                    span: Span {
                                        start: 81,
                                        end: 84,
                                    },
                                    value: 1.2,
                                    raw: "1.2",
                                },
                            ),
                        },
                    ),
                ),
            ],
        ),
    },
    errors: [],
}
